Replace get_special etc. with new attribute system

fingolfin opened the pull request:
The new system allows attaching attributes to any mutable struct types. There are two means to achieve that (replacing the old `@declare_other` attribute which for now is still supported for backwards compatibility):

1. by prefixing a mutable struct type declaration with `@attributes`, resulting in internal storage of attributes
2. by applying `@attributes` to the name of a mutable struct type, for external storage (useful if you cannot directly modify a type definition, e.g. because it is in another package)

In addition, attribute storage is supported for singleton types via external storage.

The system is designed so that one can plug in completely different attribute storage systems if so desired.

- deprecate `get_special` in favor of `get_attribute`
- deprecate `set_special` in favor of `set_attribute!`
- deprecate `hasspecial` in favor of `_has_attributes!`
- rename `get_special!` to `get_attribute!` (nothing used it yet, so no `@deprecate` for this one)
- add a `get_attribute` method taking a default value
- add `set_attribute!(G, key, val)` to match the corresponding `get_attribute` / `get_attribute!` methods
- fix bug in `get_attribute!` (formerly known as `get_special!`) that made it do the wrong thing if `G.other` was not yet set

See also https://github.com/oscar-system/Oscar.jl/issues/843

This is incomplete, but I wanted to post it now to make it visible and perhaps get some early feedback. Things left to do include:
- [x] fill in docstrings
- [x] verify this really doesn't break existing code
- [x] consider replacing `@declare_other` by a new macro that behaves more like `Base.@kwdef`, i.e., you write `@attributes mutable struct ... end` and it injects the declaration there. This has the advantage of being more flexible; e.g. we could treat `struct` and `mutable struct` differently and do other cool tricks.

fingolfin commented:
@thofma @fieker Of course I need to add in docstrings, etc., as described above; but feel free to suggest / request further changes already now.

thofma commented:
I am happy with the proposed changes. I am also fine with pulling the `@atttributes` before the `struct`.

fingolfin commented:
Made some major changes and improvements. I still need to fill in a few docstrings and resolve one TODO in the `@attributes` macro, but the rest could already be reviewed (ping @fieker)

thofma commented on src/Attributes.jl:
```diff
@@ -0,0 +1,210 @@
+julia> set_attribute!(G, :isfinite, :true)
+
+julia> get_attribute(G, :isfinite)
+42
```
```suggestion
true
```

thofma commented on src/Attributes.jl:
```diff
@@ -0,0 +1,210 @@
+julia> set_attribute!(G, :isfinite, :true)
+
+julia> get_attribute(G, :isfinite)
+42
```
```suggestion
true
```

thofma commented on src/Attributes.jl:
```diff
@@ -0,0 +1,210 @@
+function set_attribute!(G::Any, data::Pair{Symbol, <:Any}...)
+   D = _get_attributes!(G)
+   for d in data
+      push!(D, d)
```
Is `D` a dictionary?

fingolfin replied:
Yes. This code is copied from the old `set_special`

thofma reviewed (commented):
I like it. Left a few remarks/questions.

fingolfin commented:
This is now ready for merging from my POV.

thofma commented:
OK, so the plan is to keep this a breaking change, tag a new version and then adjust downstream? Fine with me, just wanted to confirm.

fingolfin commented:
Ideally I'd like to make this non-breaking. I'll investigate the breakage later and try to come up with a fix. Also, I noticed a few typos and perhaps I can add support for singleton types, too.

fingolfin commented:
Hmm, it fails because  we use `set_special` and `get_special` on `AnticNumberField` and there are custom methods of those for `AnticNumberField`. I thought our use of `@deprecate` would mesh OK with this but apparently not. I'll investigate.

In the meantime, I added support for singleton types and tweaked some comments etc.

fingolfin commented:
OK, now I understand: the problem is that AA itself uses attributes for `set_name!`, `show_name` etc. -- and I changed those to use `set_attribute!` and `get_attribute`. But that then doesn't pick up the custom `set_special` and `get_special` methods.

So for now I added a commit which reverts the `set_name` code to use `set_special`/`get_special` again. We can revert it in the future.

fingolfin commented:
Tests pass now :-)

fingolfin commented:
@thofma @wbhart @fieker can this be merged and a new AA release be made (soon, at least)?
